### PR TITLE
Init state without configs

### LINKED ISSUE #51: Initializing state without configs
### Description
[//]: # (Provide a clear and concise description of the issue.)

Currently working on initializing the simulation state without configs (so configs only stay on client side). 

Combined with issue #54, it would enable @Marsolo1 to easily create scenes with custom agents positions ...

### PR DESCRIPTION
**Pull request by corentinlger**
## Description

The PR modifies a lot of lines in the code, but there aren't tricky logic changes, we just modify the way to initialize the state without configs. I'll try to be as clear as I can when explaining the modifications.

I moved all the  state classes and definitions from ```sim_computation.py``` to a new file called ```states.py```.

The modifications also include refactoring imports in multiple files to align with the new file structure, enhancing readability and preventing linting errors.

In states.py, five functions have been implemented to initialize various states, such as simulator_state, nve_state, agent_state, objects_state, and the combined state. These changes offer greater flexibility in initializing states, allowing for customization such as specifying agent positions or using random values. Future iterations will extend this approach to more attributes, such as colors, diameters, behaviors ... to provide even more customization options.

Additionally, default values have been defined within these classes, and type hints have been added to functions for clarity. But I can still add some more docstrings to define what each parameter does.

## Related Issue (if applicable)

closes #51 

## How to Test
Launch the server
```
python3 scripts/run_server.py
```
Launch the Panel interface
```
panel serve scripts/run_interface.py --autoreload
```

Check if nothing is broken with these two lines, normally it should be the case because only code changing the state init has been modified.

Normally I added some ugly lines to show that you can initiate the position of agents with a list in run_server. Run :

```
python3 scripts/run_server.py --custom_pos
```

You can also try to launch the simulation with a custom number of agents or objects (will be usefull later for eco-evo simulations). Actually for this feature I think it would make more sense to rename the current ```n_agents``` to ```max_agents```, and ```n_existing_agents``` to ```n_agents``` (same for objects).

```
python3 scripts/run_server.py --n_existing_agents 7 --n_existing_objects 1
```

And relaunch the interface, check that the agents are indeed defined on the diagonal of the box. Later, this will easiely enable to define custom scenes, notably when using external files to store th epositions, colors ... as mentionned in #54

### CODE REVIEW

File: scripts/run_server.py [hunk 1]
```diff
@@ -35,34 +38,46 @@ def parse_args():
...
+        coord = 0
+        positions = []
+        # Here all the agents are placed on the box diagonal
+        for i in range(args.n_agents):
+            coord += 10
+            positions.append([coord, coord])
+    else:
+        positions = None
+
+    simulator_state = init_simulator_state(
```
Comment (Marsolo1) [inline scripts/run_server.py:52]: This part works to show that we can use a list of coordinates, but I think it should be removed before the merge

**Review by Marsolo1 — APPROVED** [review]
The new system seems to work fine for me, when this PR is merged i'll add the Hydra config files so that we can finally have scene files

File: scripts/run_server.py [hunk 2]
```diff
@@ -1,23 +1,26 @@
...
 lg = logging.getLogger(__name__)
 
 def parse_args():
     parser = argparse.ArgumentParser(description='Simulator Configuration')
+    parser.add_argument('--custom_pos', action='store_true', help='Just a test arg to wether use custom or random pos')
     parser.add_argument('--box_size', type=float, default=100.0, help='Size of the simulation box')
     parser.add_argument('--n_agents', type=int, default=10, help='Number of agents')
+    parser.add_argument('--n_existing_agents', type=int, default=10, help='Number of agents')
     parser.add_argument('--n_objects', type=int, default=2, help='Number of objects')
+    parser.add_argument('--n_existing_objects', type=int, default=2, help='Number of agents')
```
Comment (clement-moulin-frier) [inline scripts/run_server.py:23]: I agree with your proposition that " it would make more sense to rename the current n_agents to max_agents, and n_existing_agents to n_agents (same for objects)."

File: scripts/run_server.py [hunk 3]
```diff
@@ -1,23 +1,26 @@
...
 from vivarium.simulator.simulator import Simulator
 from vivarium.simulator.sim_computation import dynamics_rigid
-from vivarium.controllers.converters import set_state_from_config_dict
 from vivarium.simulator.grpc_server.simulator_server import serve
 
 lg = logging.getLogger(__name__)
 
 def parse_args():
     parser = argparse.ArgumentParser(description='Simulator Configuration')
+    parser.add_argument('--custom_pos', action='store_true', help='Just a test arg to wether use custom or random pos')
```
[the remark this answers is not in the document]
Comment (clement-moulin-frier) [reply scripts/run_server.py:18]: EDIT: I see in your reply to Martial below that you plan to remove it, all good

File: scripts/run_server.py (hunk 2 above)
Comment (clement-moulin-frier) [reply scripts/run_server.py:23]: Not sure if you already do it, but it will be useful to add an `assert` checking that the number of existing agents/objects does not exceed the max number

File: tests/test_simulator_init.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    prox_cos_min = 0.0
+    color = "red"
+
+    simulator_state = init_simulator_state(
+        box_size=box_size,
+          n_agents=n_agents,
+          n_objects=n_objects)
+
+    nve_state = init_nve_state(
+        simulator_state,
```
Comment (clement-moulin-frier) [inline tests/test_simulator_init.py:52]: wrong indentation

File: tests/test_simulator_init.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        nve_state=nve_state
+        )
+
+    simulator = Simulator(state, behaviors.behavior_bank, dynamics_rigid)
+
+    assert simulator
+
+
+def test_init_simulator_args():
+    """ Test the initialization of state with arguments """
```
[the remark this answers is not in the document]
Comment (clement-moulin-frier) [reply tests/test_simulator_init.py:31]: Another useful test to do would be to check if (some of) the values in the created state are indeed the ones that are provided in the args or defaults (e.g. just adding a few `assert` in all your test functions)

File: vivarium/simulator/states.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
+from typing import Optional, List, Union
+from enum import Enum
+
+import matplotlib.colors as mcolors
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:4]: Not related to this PR, but I just notice that this requires `matplotlib`. We should either:

- Add matplotlib to the `requirements.txt` (if not already there)
- Or: find a way to do the same thing without using matlplotlib. The `mcolors` function is I think just used to convert colors between an vector RGB format (used in State) and a string format (used in the config). Maybe there is a way to do this without relying on matlplotlib.

If possible, I think the second option is better, because imposing to install matlplotlib (which is quite a big lib) just for this. We can discuss it and do it later though, you can just create a new issue atm

Comment (clement-moulin-frier) [reply vivarium/simulator/states.py:4]: There is actually a deeper issue here, see my other comment below

File: vivarium/simulator/states.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+    color: util.Array
+
+
+@dataclass
+class ObjectState:
+    nve_idx: util.Array  # idx in NVEState
+    color: util.Array
+
+
+# TODO : I think it would make more sense to have max_agents, max_objects here instead of n_***
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:69]: Agreed

File: vivarium/simulator/states.py [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+                return value[self.e_cond(e_type)]
+        return wrapper
+
+
+# Helper function to transform a color string into rgb with matplotlib colors
+def _string_to_rgb(color_str):
+    return jnp.array(list(mcolors.to_rgb(color_str)))
+
+
+def init_simulator_state(
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:148]: We can do this in another Issue/PR, but it would be better if default values are defined at a single place in the code. For now we have them in several places, e.g. here, in the run scripts, in the Config classes (and maybe elsewhere), this will be very hard to maintain. We could e.g. have them in a dedicated file and all the classes/functions that need them would import them from there. Might be easier to do it in the context of #30 though (if you agree please update #30 or related PR so that we don't forget, otherwise let's discuss it this afternoon)

File: vivarium/simulator/states.py [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+    assert (len(positions) == n_elements if positions else True)
+    # If positions are passed, transform them in jax array
+    if positions:
+        positions = jnp.array(positions)
+    # Else initialize random positions
+    else:
+        positions = random.uniform(key_pos, (n_elements, n_dims)) * box_size
+    return positions
+
+def _init_existing(n_existing, n_elements):
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:186]: I am not sure what an "element" refers to here. Isn't it instead related to an entity? In that case change it to e.g. `n_entities`, it will be more clear (in general we should avoid multiplying the terms we use)

Comment (clement-moulin-frier) [reply vivarium/simulator/states.py:186]: If you do it report it everywhere you use '"element"

File: vivarium/simulator/states.py [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        non_existing_arr = jnp.zeros((n_elements - n_existing))
+        exists_array = jnp.concatenate((existing_arr, non_existing_arr))
+    else:
+        exists_array = jnp.ones((n_elements))
+    return exists_array
+
+# TODO : Should also add union float, list[float] for friction, diameter ..
+# Also think it would be easier to only handle the nve state in the physics engine. (if it doesn't make the  simulation run slower)
+# Here it makes a really long function, and it isn't very modular 
+# Plus we store different attributes of entities (e.g Agents positions and sensors) in different dataclasses
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:199]: Let's discuss it, as we said last time it is part of a more general issue about how we handle conversions between data types that encode the simular infomation (AgentState, NVEState, Config, proto ..)

Comment (clement-moulin-frier) [reply vivarium/simulator/states.py:199]: You can create an issue about explaining what we have already discussed on this topic (a potential solution would be to have useful helper functions in the State class)

File: vivarium/simulator/states.py [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+        seed: int = 0,
+        ) -> NVEState:
+    """
+    Initialize nve state with given parameters
+    """
+    n_agents = simulator_state.n_agents[0]
+    n_objects = simulator_state.n_objects[0]
+    n_entities = n_agents + n_objects
+
+    key = random.PRNGKey(seed)
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:219]: I'm not sure this follows the good practice about how to deal with randomness in JAX, see https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html

Typically, the seed should be defined in a single place in the code, and all functions requiring a random number should use `jax.random.split`. `PRNGKey` should be called only once in the entire code base. (it's how I understand it at least) 

See if you can fix it here, otherwise create an issue (point at the code here so that we don't forget it)

File: scripts/run_server.py (hunk 2 above)
Comment (corentinlger) [reply scripts/run_server.py:23]: Yes I already check this

File: tests/test_simulator_init.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (corentinlger) [reply tests/test_simulator_init.py:31]: Indeed I'll add that

File: vivarium/simulator/states.py (hunk 6 above)
Comment (corentinlger) [reply vivarium/simulator/states.py:4]: I agree with you only using matplotlib for this isn't really useful. But it is needed to launch the interface with panel anyway I am not wrong. That's why it was already in the requirements.txt.

File: vivarium/simulator/states.py [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+        dt=jnp.array([dt], dtype=float),
+        freq=jnp.array([freq], dtype=float),
+        neighbor_radius=jnp.array([neighbor_radius], dtype=float),
+        # Use 1*bool to transform True to 1 and False to 0
+        to_jit= jnp.array([1*to_jit]),
+        use_fori_loop=jnp.array([1*use_fori_loop]))
+
+
+def _init_positions(key_pos, positions, n_elements, box_size, n_dims=2):
+    assert (len(positions) == n_elements if positions else True)
```
[the remark this answers is not in the document]
Comment (corentinlger) [reply vivarium/simulator/states.py:177]: Agreed !

File: vivarium/simulator/states.py (hunk 11 above)
Comment (corentinlger) [reply vivarium/simulator/states.py:219]: It is a good question. I am not sure about this (because in our case this key only serves to initialize random positions, and they key in simulator serve other purposes) but it will be easy to change if we want to. Actually I checked some RL repos in Jax and they seem to define keys with seeds in different parts of the code (ex: in environments and in algorithms). 

From what I saw we only have another key initialization in state, we can just change the init methods so it doesn't accept a seed that generates a key, but directly a key (that you generate at the  beginning of the script you run).

**Comment by Marsolo1** [comment]
>     When running the server with `--n_existing_agents 7 --n_existing_objects 1`, then selecting a non-existing agent and object and activating the the `visible` checkbox, the selected entity appear for a very short time and disappear right away. I'm not sure it is specific to this PR though, it might be a more general bug. Can you have a look @Marsolo1 ?

Yes I already saw this bug and fixed it but it was in another PR, so I just pushed the fix to main directly (it was only a False in the wrong place). If we merge this branch with main it should fix it here too

File: vivarium/simulator/states.py [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+# TODO : Should also add union float, list[float] for friction, diameter ..
+# Also think it would be easier to only handle the nve state in the physics engine. (if it doesn't make the  simulation run slower)
+# Here it makes a really long function, and it isn't very modular 
+# Plus we store different attributes of entities (e.g Agents positions and sensors) in different dataclasses
+def init_nve_state(
+        simulator_state: SimulatorState,
+        diameter: float = 5.,
+        friction: float = 0.1,
+        mass_center: float = 1.,
+        mass_orientation: float = 0.125,
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:205]: We should be able to pass a full array as well here, e.g. if we want to set different diameters for different entities. A solution could be to check the type of this argument : if it is a scalar, all entities will have the same value, if it is an array, we set them as such. Will require to explain it in a docstring ;)

File: vivarium/simulator/states.py [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+        mass=RigidBody(center=jnp.full((n_entities, 1), mass_center), orientation=jnp.full((n_entities), mass_orientation)),
+        entity_type=entity_types,
+        entity_idx = jnp.array(list(range(n_agents)) + list(range(n_objects))),
+        diameter=jnp.full((n_entities), diameter),
+        friction=jnp.full((n_entities), friction),
+        exists=exists
+        )
+
+
+# Could implement it as a static or class method
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:254]: It's true that this function and the others could instead be static/class methods of their associated State class. Let's discuss it

File: vivarium/simulator/states.py [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+def init_agent_state(
+        simulator_state: SimulatorState,
+        behavior: int = 1,
+        wheel_diameter: float = 2.,
+        speed_mul: float = 1.,
+        theta_mul: float = 1.,
+        prox_dist_max: float = 40.,
+        prox_cos_min: float = 0.,
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:262]: Idem, we should be able to pass arrays if we want to

File: vivarium/simulator/states.py [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+        color: str = "blue"
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:263]: To be discussed, I'm not sure it is super coherent to pass colors as strings, since we are in the simulator world here. For example, `behavior` above is defined as an int, not a string (whereas it is defined as a string in the client Configs, as for colors). 
Also, specifying colors as a RGB vector here will solve the issue I mentioned about matlplotlib above. Maybe the best solution would be to have a RGB <-> String converter in a utils module, that can be easily imported wherever we need this. For example here ; or in the run script

File: vivarium/simulator/states.py [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+        ) -> AgentState:
+    """
+    Initialize agent state with given parameters
+    """
+    n_agents = simulator_state.n_agents[0]
+    # TODO : Allow to define custom list of behaviors, wheel_diameters ... (in fact for all parameters)
+    # if the shape if just 1 value, assign it to all agents
+    # else, ensure you are given a list of arguments of size n_agents and transform it into a jax array
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:271]: Great, we thought of the same thing :) I agree, let's do this

File: vivarium/simulator/states.py [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+        theta_mul=jnp.full((n_agents), theta_mul),
+        proxs_dist_max=jnp.full((n_agents), prox_dist_max),
+        proxs_cos_min=jnp.full((n_agents), prox_cos_min),
+        color=jnp.tile(_string_to_rgb(color), (n_agents, 1))
+    )
+
+
+def init_object_state(
+        simulator_state: SimulatorState,
+        color: str = "red"
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:289]: Same comment as above for color

File: vivarium/simulator/states.py [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+        nve_state: NVEState
+        ) -> State:
+  
+    return State(
+        simulator_state=simulator_state,
+        agent_state=agents_state,
+        object_state=objects_state,
+        nve_state=nve_state
+    )
+
```
Comment (clement-moulin-frier) [inline vivarium/simulator/states.py:316]: A general comment to discuss:

If I understand well, with this PR a user will be able to define a custom state more easily from a run script, which is nice indeed. But if the objective is to simplify the life of the use, I am wondering if we shouldn't abstract the distinction between the NVEState and the Agent/ObjectState. How about to just having a function where a user can just enter what he wants (e.g. how many agents/objects, which positions etc ...) and then the function will deal with how to construct the State, hiding this complexity to the user.

**Review by clement-moulin-frier — COMMENTED** [review]
This is the second part of my review, which is now completed. Please read the comments and then we discuss

File: vivarium/simulator/states.py [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+    """
+    Initialize nve state with given parameters
+    """
+    n_agents = simulator_state.n_agents[0]
+    n_objects = simulator_state.n_objects[0]
+    n_entities = n_agents + n_objects
+
+    key = random.PRNGKey(seed)
+    key_pos, key_or = random.split(key)
+    key_ag, key_obj = random.split(key_pos)
```
[the remark this answers is not in the document]
Comment (corentinlger) [reply vivarium/simulator/states.py:221]: you prefer to do a random.split(key, 4) ?
